File: day3/partitioner.py
import math
import numpy as np

class Partitioner:
    """2-dimensional domain decomposition of a 3-dimensional computational
       grid among MPI ranks on a communicator."""
    
    
    def __init__(self, comm, domain, num_halo, periodic = (True, True)):
        assert len(domain) == 3, \
            "Must specify a three-dimensional domain"
        assert domain[0] > 0 and domain[1] > 0 and domain[2] > 0, \
            "Invalid domain specification (negative size)"
        assert num_halo >= 0, "Number of halo points must be zero or positive"

        self.__comm = comm
        self.__num_halo = num_halo
        self.__periodic = periodic

        self.__rank = comm.Get_rank()
        self.__num_ranks = comm.Get_size()
        self.__global_shape = [domain[0], domain[1] + 2 * num_halo, domain[2] + 2 * num_halo]

        size = self.__setup_grid()
        assert domain[1] >= size[0] and domain[2] >= size[1], "Domain is too small for number of ranks"
        self.__setup_domain(domain, num_halo)
        
    def comm(self):
        """Returns the MPI communicator use to setup this partitioner"""
        return self.__comm
    
    
    def num_halo(self):
        """Returns the number of halo points"""
        return self.__num_halo
    
    
    def periodic(self, dim=None):
        """Returns the periodicity of individual or all dimensions"""
        if dim is not None:
            return self.__periodic[dim]
        else:
            return self.__periodic

        
    def rank(self):
        """Returns the rank of the current MPI worker"""
        return self.__rank
    
    
    def num_ranks(self):
        """Returns the number of ranks that have been distributed by this partitioner"""
        return self.__num_ranks
    
        
    def shape(self):
        """Returns the shape of a local field (including halo points)"""
        return self.__shape
    
    
    def global_shape(self):
        """Returns the shape of a local field (including halo points)"""
        return self.__global_shape
    

    def grid_size(self):
        """Dimensions of the two-dimensional worker grid"""
        return self.__grid_size
    
    
    def grid_position(self):
        """Position of current rank on two-dimensional worker grid"""
        return self.__rank_to_position(self.__rank)


    def left(self):
        """Returns the rank of the left neighbor"""
        return self.__get_neighbor_rank( [0, -1] )
    
    
    def right(self):
        """Returns the rank of the left neighbor"""
        return self.__get_neighbor_rank( [0, +1] )
    
    
    def top(self):
        """Returns the rank of the left neighbor"""
        return self.__get_neighbor_rank( [+1, 0] )
    
    
    def bottom(self):
        """Returns the rank of the left neighbor"""
        return self.__get_neighbor_rank( [-1, 0] )
    
    
    def scatter(self, field, root=0):
        """Scatter a global field from a root rank to the workers"""
        if self.__rank == root:
            assert np.any(field.shape[0] == np.array(self.__global_shape[0])), \
                "Field does not have correct shape"
        assert 0 <= root < self.__num_ranks, "Root processor must be a valid rank"
        if self.__num_ranks == 1:
            return field
        sendbuf = None
        if self.__rank == root:
            sendbuf = np.empty( [self.__num_ranks,] + self.__max_shape, dtype=field.dtype )
            for rank in range(self.__num_ranks):
                j_start, i_start, j_end, i_end = self.__domains[rank]
                sendbuf[rank, :, :j_end-j_start, :i_end-i_start] = field[:, j_start:j_end, i_start:i_end]
        recvbuf = np.empty(self.__max_shape, dtype=field.dtype)
        self.__comm.Scatter(sendbuf, recvbuf, root=root)
        j_start, i_start, j_end, i_end = self.__domain
        return recvbuf[:, :j_end-j_start, :i_end-i_start].copy()
        
    
    def gather(self, field, root=0):
        """Gather a distributed fields from workers to a single global field on a root rank"""
        assert np.any(field.shape == np.array(self.__shape)), "Field does not have correct shape"
        assert -1 <= root < self.__num_ranks, "Root processor must be -1 (all) or a valid rank"
        if self.__num_ranks == 1:
            return field
        j_start, i_start, j_end, i_end = self.__domain
        sendbuf = np.empty( self.__max_shape, dtype=field.dtype )
        sendbuf[:, :j_end-j_start, :i_end-i_start] = field
        recvbuf = None
        if self.__rank == root or root == -1:
            recvbuf = np.empty( [self.__num_ranks,] + self.__max_shape, dtype=field.dtype )
        if root > -1:
            self.__comm.Gather(sendbuf, recvbuf, root=root)
        else:
            self.__comm.Allgather(sendbuf, recvbuf)
        global_field = None
        if self.__rank == root or root == -1:
            global_field = np.empty(self.__global_shape, dtype=field.dtype)
            for rank in range(self.__num_ranks):
                j_start, i_start, j_end, i_end = self.__domains[rank]
                global_field[:, j_start:j_end, i_start:i_end] = recvbuf[rank, :, :j_end-j_start, :i_end-i_start]
        return global_field
                
    
    def compute_domain(self):
        """Return position of subdomain without halo on the global domain"""
        return [self.__domain[0] + self.__num_halo, self.__domain[1] + self.__num_halo, \
                self.__domain[2] - self.__num_halo, self.__domain[3] - self.__num_halo]


    def __setup_grid(self):
        """Distribute ranks onto a Cartesian grid of workers"""
        for ranks_x in range(math.floor( math.sqrt(self.__num_ranks) ), 0, -1):
            if self.__num_ranks % ranks_x == 0:
                break
        self.__grid_size = (self.__num_ranks // ranks_x, ranks_x)
        return self.__grid_size
        

    def __get_neighbor_rank(self, offset):
        """Get the rank ID of a neighboring rank at a certain offset relative to the current rank"""
        position = self.__rank_to_position(self.__rank)
        pos_y = self.__cyclic_offset(position[0], offset[0], self.__grid_size[0], self.__periodic[0])
        pos_x = self.__cyclic_offset(position[1], offset[1], self.__grid_size[1], self.__periodic[1])
        return self.__position_to_rank( [pos_y, pos_x] )


    def __cyclic_offset(self, position, offset, size, periodic=True):
        """Add offset with cyclic boundary conditions"""
        pos = position + offset
        if periodic:
            while pos < 0:
                pos += size
            while pos > size - 1:
                pos -= size
        return pos if -1 < pos < size else None


    def __setup_domain(self, shape, num_halo):
        """Distribute the points of the computational grid onto the Cartesian grid of workers"""
        assert len(shape) == 3, "Must pass a 3-dimensional shape"
        size_z = shape[0]
        size_y = self.__distribute_to_bins(shape[1], self.__grid_size[0])
        size_x = self.__distribute_to_bins(shape[2], self.__grid_size[1])

        pos_y = self.__cumsum(size_y, initial_value=num_halo)
        pos_x = self.__cumsum(size_x, initial_value=num_halo)

        domains = []
        shapes = []
        for rank in range(self.__num_ranks):
            pos = self.__rank_to_position(rank)
            domains += [[ pos_y[pos[0]] - num_halo, pos_x[pos[1]] - num_halo, \
                          pos_y[pos[0] + 1] + num_halo, pos_x[pos[1] + 1] + num_halo ]]
            shapes += [[ size_z, domains[rank][2] - domains[rank][0], \
                                 domains[rank][3] - domains[rank][1] ]]
        self.__domains, self.__shapes =  domains, shapes
        
        self.__domain, self.__shape = domains[self.__rank], shapes[self.__rank]
        self.__max_shape = self.__find_max_shape( self.__shapes )


    def __distribute_to_bins(self, number, bins):
        """Distribute a number of elements to a number of bins"""
        n = number // bins
        bin_size = [n] * bins
        # make bins in the middle slightly larger
        extend = number - n * bins
        if extend > 0:
            start_extend = bins // 2 - extend // 2
            bin_size[start_extend:start_extend + extend] = \
                [ n + 1 for n in bin_size[start_extend:start_extend + extend] ]
        return bin_size

    
    def __cumsum(self, array, initial_value=0):
        """Cumulative sum with an optional initial value (default is zero)"""
        cumsum = [initial_value]
        for i in array:
            cumsum += [ cumsum[-1] + i ]
        return cumsum


    def __find_max_shape(self, shapes):
        max_shape = shapes[0]
        for shape in shapes[1:]:
            max_shape = list(map(max, zip(max_shape, shape)))
        return max_shape
    

    def __rank_to_position(self, rank):
        """Find position of rank on worker grid"""
        return ( rank // self.__grid_size[1], rank % self.__grid_size[1] )
    

    def __position_to_rank(self, position):
        """Find rank given a position on the worker grid"""
        if position[0] is None or position[1] is None:
            return None
        else:
            return position[0] * self.__grid_size[1] + position[1]
    

File: projects/2024/project09_halo_update_performance/partitioner.cpp
// Author: Philip Ploner
// Date: 2024-08-24
// 2-dimensional domain decomposition of a 3-dimensional computational grid among MPI ranks on a communicator, implemented in C++
// This code follows the same structure as the partitioner.py file from HPC4WC/day3/


#include <cassert>
#include <vector>
#include <cmath>
#include <memory>
#include <thread>
#include <chrono>
#include <mpi.h>
#include "utils.h"

class Partitioner
{
public:
    Partitioner(MPI_Comm comm, std::vector<int> domain, int num_halo, std::pair<bool, bool> periodic = {true, true})
    {
        assert(domain.size() == 3);
        assert(domain[0] > 0 && domain[1] > 0 && domain[2] > 0);
        assert(num_halo >= 0);

        comm_ = comm;
        num_halo_ = num_halo;
        periodic_ = periodic;

        xsize_ = domain[2] + 2*num_halo;
        ysize_ = domain[1] + 2*num_halo;
        zsize_ = domain[0];

        MPI_Comm_rank(comm_, &rank_);
        MPI_Comm_size(comm_, &num_ranks_);
        global_shape_ = {domain[0], domain[1] + 2 * num_halo, domain[2] + 2 * num_halo};

        size_ = setup_grid(); //[2,1], domain: [4,20,20]
        assert(domain[1] >= size_[0] && domain[2] >= size_[1]);
        setup_domain(domain, num_halo);
    }

    // Returns the MPI communicator use to setup this partitioner
    MPI_Comm comm() const
    {
        return comm_;
    }

    // Returns the number of halo points
    int num_halo() const
    {
        return num_halo_;
    }

    // Returns the periodicity of individual or all dimensions
    std::pair<bool, bool> periodic() const
    {
        return periodic_;
    }

    // Returns the rank of the current MPI worker
    int rank() const
    {
        return rank_;
    }

    // Returns the number of ranks that have been distributed by this partitioner
    int num_ranks() const
    {
        return num_ranks_;
    }

    // "Returns the shape of a local field (including halo points)
    std::vector<int> shape() const
    {
        return shape_;
    }

    // Returns the shape of a global field (including halo points)
    std::vector<int> global_shape() const
    {
        return global_shape_;
    }

    // Dimensions of the two-dimensional worker grid
    std::vector<int> size() const
    {
        return size_;
    }

    // Position of current rank on two-dimensional worker grid
    std::pair<int, int> position() 
    {
        return rank_to_position(rank_);
    }

    // Get the rank ID of a neighboring rank at a certain offset relative to the current rank
    int get_neighbor_rank(std::pair<int, int> offset) 
    {
        auto position = rank_to_position(rank_);
        auto pos_y = cyclic_offset(position.first, offset.first, size_[0], periodic_.first);
        auto pos_x = cyclic_offset(position.second, offset.second, size_[1], periodic_.second);
        return position_to_rank({pos_y, pos_x});
    }

    // Returns the rank of the left neighbor
    int left() 
    {
        return get_neighbor_rank({0, -1});
    }

    // Returns the rank of the right neighbor
    int right() 
    {
        return get_neighbor_rank({0, 1});
    }

    // Returns the rank of the top neighbor
    int top() 
    {
        return get_neighbor_rank({1, 0});
    }

    // Returns the rank of the bottom neighbor
    int bottom() 
    {
        return get_neighbor_rank({-1, 0});
    }

    // Scatter a global field from a root rank to the workers
    Storage3D<double> scatter(Storage3D<double> &field, int root = 0)
    {
        if (rank_ == root)
        {
            assert(field.zMax() == global_shape_[0]);
        }
        assert(0 <= root && root < num_ranks_);
        if (num_ranks_ == 1)
        {
            return field;
        }
        std::unique_ptr<std::vector<double>> sendbuf;

        int size_of_rank = max_shape_[2] * max_shape_[1] * max_shape_[0];
        if (rank_ == root)
        {
            sendbuf = std::make_unique<std::vector<double>>(num_ranks_ * max_shape_[2] * max_shape_[1] * max_shape_[0], 0.0);
            for (int rank = 0; rank < num_ranks_; ++rank)
            {
                int j_start = domains_[rank][0], i_start=domains_[rank][1], j_end=domains_[rank][2], i_end=domains_[rank][3];
                for (int k = 0; k < max_shape_[0]; ++k)
                {
                    for (int j = j_start; j < j_end; ++j)
                    {
                        for (int i = i_start; i < i_end; ++i)
                        {
                            (*sendbuf)[rank*size_of_rank + (i - i_start) + (j - j_start) * (i_end - i_start) + k * (i_end - i_start) * (j_end - j_start)] = field(i,j,k);
                        }
                    }
                }
            }
        }

        std::vector<double> recvbuf(max_shape_[2] * max_shape_[1] * max_shape_[0], 0.0);
        MPI_Barrier(comm_);
        MPI_Scatter(rank_ == root ? &(*sendbuf)[0] : nullptr, max_shape_[0] * max_shape_[1] * max_shape_[2], 
        MPI_DOUBLE, &recvbuf[0], max_shape_[0] * max_shape_[1] * max_shape_[2], MPI_DOUBLE, root, comm_);
        MPI_Barrier(comm_);


        int j_start=domain_[0], i_start=domain_[1], j_end=domain_[2], i_end=domain_[3];
        static Storage3D<double> recvbuf_3d (max_shape_[2] - 2*num_halo_, max_shape_[1] - 2*num_halo_, max_shape_[0], num_halo_, 0.0);
        for (int k = 0; k < max_shape_[0]; ++k)
        {
            for (int j = 0; j < j_end - j_start; ++j)
            {
                for (int i = 0; i < i_end - i_start; ++i)
                {
                    recvbuf_3d(i,j,k) = recvbuf[i + j * (i_end - i_start) + k * (i_end - i_start) * (j_end - j_start)];
                }
            }
        }

        MPI_Barrier(comm_);
        return recvbuf_3d;
    }

    // Gather a distributed fields from workers to a single global field on a root rank
    std::unique_ptr<Storage3D<double>> gather(Storage3D<double> &field, int root = 0)
    {
        assert(field.zMax() == shape_[0]);
        assert(-1 <= root && root < num_ranks_);
        if (num_ranks_ == 1)
        {
            return std::make_unique<Storage3D<double>>(field);
        }
        int j_start=domain_[0], i_start=domain_[1], j_end=domain_[2], i_end=domain_[3];
        std::vector<double> sendbuf (max_shape_[2] * max_shape_[1] * max_shape_[0], 0.0);
        for (int k = 0; k < shape_[0]; ++k)
        {
            for (int j = j_start; j < j_end; ++j)
            {
                for (int i = i_start; i < i_end; ++i)
                {
                    sendbuf[(i - i_start) + (j - j_start) * (i_end - i_start) + k * (i_end - i_start) * (j_end - j_start)] = field(i - i_start, j - j_start, k);
                }
            }
        }
        MPI_Barrier(comm_);
        std::unique_ptr<std::vector<double>> recvbuf;
        if (rank_ == root || root == -1)
        {
            recvbuf = std::make_unique<std::vector<double>>(num_ranks_ * max_shape_[2] * max_shape_[1] * max_shape_[0], 0.0);
        }

        MPI_Barrier(comm_);
        if (root > -1)
        {
            MPI_Gather(&sendbuf[0], max_shape_[0] * max_shape_[1] * max_shape_[2], MPI_DOUBLE, rank_ == root ? &(*recvbuf)[0] : nullptr, max_shape_[0] * max_shape_[1] * max_shape_[2], MPI_DOUBLE, root, comm_);
        }
        else
        {
            MPI_Allgather(&sendbuf[0], max_shape_[0] * max_shape_[1] * max_shape_[2], MPI_DOUBLE, rank_ == root ? &(*recvbuf)[0] : nullptr, max_shape_[0] * max_shape_[1] * max_shape_[2], MPI_DOUBLE, comm_);
        }
        MPI_Barrier(comm_);
        int size_of_rank = max_shape_[2] * max_shape_[1] * max_shape_[0];
        if (rank_ == root || root == -1)
        {
            std::unique_ptr<Storage3D<double>> global_field = std::make_unique<Storage3D<double>>(global_shape_[2] - 2*num_halo_, global_shape_[1] - 2*num_halo_, global_shape_[0], num_halo_, 0.0);
            for (int rank = 0; rank < num_ranks_; ++rank)
            {
                int j_start = domains_[rank][0], i_start=domains_[rank][1], j_end=domains_[rank][2], i_end=domains_[rank][3];
                for (int k = 0; k < max_shape_[0]; ++k)
                {
                    for (int j = j_start; j < j_end; ++j)
                    {
                        for (int i = i_start; i < i_end; ++i)
                        {
                            (*global_field)(i,j,k) = (*recvbuf)[rank*size_of_rank + (i - i_start) + (j - j_start) * (i_end - i_start) + k * (i_end - i_start) * (j_end - j_start)];
                        }
                    }
                }
            }
            MPI_Barrier(comm_);
            return global_field;
        }
        else
        {
            MPI_Barrier(comm_);
            return nullptr;
        }
    }

    // Return position of subdomain without halo on the global domain
    std::vector<int> compute_domain() const
    {
        return {domain_[0] + num_halo_, domain_[1] + num_halo_, domain_[2] - num_halo_, domain_[3] - num_halo_};
    }

private:
    MPI_Comm comm_;
    int num_halo_;
    int xsize_;
    int ysize_;
    int zsize_;
    std::pair<bool, bool> periodic_;
    int rank_;
    int num_ranks_;
    std::vector<int> global_shape_;
    std::vector<int> size_;
    std::vector<std::vector<int>> domains_;
    std::vector<std::vector<int>> shapes_;
    std::vector<int> domain_;
    std::vector<int> shape_;
    std::vector<int> max_shape_;

    // Distribute ranks onto a Cartesian grid of workers
    std::vector<int> setup_grid()
    {
        int ranks_x_size;
        for (int ranks_x = std::floor(std::sqrt(num_ranks_)); ranks_x > 0; --ranks_x)
        {
            if (num_ranks_ % ranks_x == 0)
            {
                ranks_x_size = ranks_x;
                break;
            }
        }
        return {num_ranks_ / ranks_x_size, ranks_x_size};
    }

    // Add offset with cyclic boundary conditions
    int cyclic_offset(int position, int offset, int size, bool periodic = true)
    {
        int pos = position + offset;
        if (periodic)
        {
            while (pos < 0)
            {
                pos += size;
            }
            while (pos > size - 1)
            {
                pos -= size;
            }
        }
        return -1 < pos && pos < size ? pos : -1;
    }

    // "Distribute the points of the computational grid onto the Cartesian grid of workers
    void setup_domain(std::vector<int> shape, int num_halo)
    {
        int size_z = shape[0];
        auto size_y = distribute_to_bins(shape[1], size_[0]);
        auto size_x = distribute_to_bins(shape[2], size_[1]);

        auto pos_y = cumsum(size_y, num_halo);
        auto pos_x = cumsum(size_x, num_halo);

        std::vector<std::vector<int>> domains;
        std::vector<std::vector<int>> shapes;
        for (int rank = 0; rank < num_ranks_; ++rank)
        {
            auto pos = rank_to_position(rank); 
            domains.push_back({pos_y[pos.first] - num_halo, pos_x[pos.second] - num_halo, pos_y[pos.first + 1] + num_halo, pos_x[pos.second + 1] + num_halo});
            shapes.push_back({size_z, domains[rank][2] - domains[rank][0], domains[rank][3] - domains[rank][1]});
        }
        domains_ = domains;
        shapes_ = shapes;

        domain_ = domains_[rank_]; 
        shape_ = shapes_[rank_];
        max_shape_ = find_max_shape(shapes_); 
    }

    // Distribute a number of elements to a number of bins
    std::vector<int> distribute_to_bins(int number, int bins)
    {
        int n = number / bins;
        std::vector<int> bin_size(bins, n);
        int extend = number - n * bins; 
        if (extend > 0)
        {
            int start_extend = bins / 2 - extend / 2;
            for (int i = start_extend; i < start_extend + extend; ++i)
            {
                bin_size[i] += 1;
            }
        }
        return bin_size;
    }

    // Cumulative sum with an optional initial value (default is zero)
    std::vector<int> cumsum(std::vector<int> array, int initial_value = 0)
    {
        std::vector<int> cumsum = {initial_value};
        for (int i : array)
        {
            cumsum.push_back(cumsum.back() + i); 
        }
        return cumsum;
    }

    // Find largest shape in a list of shapes
    std::vector<int> find_max_shape(const std::vector<std::vector<int>> &shapes)
    {
        if (shapes.empty())
        {
            return {};
        }

        std::vector<int> max_shape = shapes[0];
        for (size_t i = 1; i < shapes.size(); ++i)
        {
            for (size_t j = 0; j < max_shape.size(); ++j)
            {
                max_shape[j] = std::max(max_shape[j], shapes[i][j]);
            }
        }
        return max_shape;
    }

    // Find position of rank on worker grid
    std::pair<int, int> rank_to_position(int rank)
    {
        return {rank / size_[1], rank % size_[1]};
    }

    // Find rank given a position on the worker grid
    int position_to_rank(std::pair<int, int> position)
    {
        if (position.first == -1 || position.second == -1)
        {
            return -1;
        }
        else
        {
            return position.first * size_[1] + position.second;
        }
    }
};
